### PR TITLE
Wip s3 filter

### PR DESCRIPTION
**Pull request by mosayyebzadeh**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/get_obj [hunk 1]
```diff
@@ -0,0 +1,303 @@
+rgw_admin.cc:  int ret = conn->forward(dpp(), user, info, nullptr, MAX_REST_RESPONSE, &in_data, &response, null_yield);
```
Comment (alimaredia) [inline src/rgw/get_obj:1]: This file can be cut from the PR for cleanup

File: src/common/options/global.yaml.in [hunk 2]
```diff
@@ -6325,3 +6325,4 @@ options:
   default: 0
   services:
   - mgr
+
```
Comment (alimaredia) [inline src/common/options/global.yaml.in:6328]: This change can be cut from the PR for cleanup

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -4026,7 +4026,6 @@ void RGWPutObj::execute(optional_yield y)
 					 s->bucket_owner.get_id(),
 					 pdest_placement, olh_epoch, s->req_id);
   }
-
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:4029]: This change can be cut from the PR for cleanup

File: src/rgw/rgw_process.cc [hunk 4]
```diff
@@ -457,6 +457,7 @@ int process_request(rgw::sal::Store* const store,
   } else {
     ldpp_dout(s, 2) << "http status=" << s->err.http_ret << dendl;
   }
+
```
Comment (alimaredia) [inline src/rgw/rgw_process.cc:460]: This change can be cut from the PR for cleanup

File: src/rgw/rgw_sal.cc [hunk 5]
```diff
@@ -367,12 +388,22 @@ StoreManager::Config StoreManager::get_config(bool admin, CephContext* cct)
   }
 #endif
 
+  //FIXME : AMIN : uncomment below lines
```
Comment (alimaredia) [inline src/rgw/rgw_sal.cc:391]: This section can be cut from the PR for cleanup

File: src/rgw/rgw_sal_store.h [hunk 6]
```diff
@@ -399,6 +399,7 @@ class StoreWriter : public Writer {
   StoreWriter(const DoutPrefixProvider *_dpp, optional_yield y) : dpp(_dpp) {}
   virtual ~StoreWriter() = default;
 
+  //virtual int prepare(optional_yield y, uint64_t obj_size = 0) {return 0;};
```
Comment (alimaredia) [inline src/rgw/rgw_sal_store.h:402]: @mosayyebzadeh why was this commented out?

File: src/vstart.sh [hunk 7]
```diff
@@ -708,6 +708,8 @@ prepare_conf() {
         enable experimental unrecoverable data corrupting features = *
         osd_crush_chooseleaf_type = 0
         debug asok assert abort = true
+	rgw_filter = s3
```
Comment (alimaredia) [inline src/vstart.sh:711]: These changes should be removed from the PR as well

File: src/rgw/rgw_sal_filter.h [hunk 8]
```diff
@@ -862,6 +862,7 @@ class FilterWriter : public Writer {
   virtual ~FilterWriter() = default;
 
   virtual int prepare(optional_yield y) { return next->prepare(y); }
+  //virtual int prepare(optional_yield y, uint64_t obj_size = 0) { return next->prepare(y); }
```
Comment (alimaredia) [inline src/rgw/rgw_sal_filter.h:865]: Why was prepare removed here?

File: src/rgw/rgw_sal_s3.cc [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+
+
+int S3FilterObject::get_obj_state(const DoutPrefixProvider* dpp, RGWObjState **state,
+			    optional_yield y, bool follow_olh){
+  RGWObjState *astate = new RGWObjState;
+  *state = astate;
+  return 0;
+}
+
+/*
```
Comment (alimaredia) [inline src/rgw/rgw_sal_s3.cc:625]: Could you cut this last chunk of code here since it's all commented out.

File: src/rgw/rgw_sal_s3.cc [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+  source->set_obj_size(received_data.length());
+
+  return 0;
+}
+
+int S3FilterObject::S3FilterReadOp::iterate(const DoutPrefixProvider* dpp, int64_t ofs,
+					int64_t end, RGWGetDataCB* cb, optional_yield y)
+{
+  ldpp_dout(dpp, 20) << " AMIN: " << __func__ << " : " << "ofs is: " << ofs << " end is: " << end << dendl;
+  /*
```
Comment (alimaredia) [inline src/rgw/rgw_sal_s3.cc:530]: If this chunk of code isn't needed, it should be removed

**Comment by alimaredia** [comment]
@mosayyebzadeh could you please clean up the PR (see the comments I left) and then rebase it

**Comment by yuvalif** [comment]
@mosayyebzadeh can you please add more info on the design here.
for example:
* is the intent here that the RGW will be an S3 proxy?
* why filter and not backend?
* what about credentials?
* did you consider using the aws-c++-sdk?
* seems like you are using blocking calls to the HTTP manager. that won't work in real deployments
* etc.

probably good to discuss in one of the RGW refactoring meetings.

**Pull request closed without merging**